### PR TITLE
Add support for OVN SB-DB Chassis_Private table

### PR DESCRIPTION
**Pull request by Pardhakeswar**
- ChassisPrivateDel() - Delete row in chassis_private table
    with given chassis name
 - ChassisPrivateList() - List all chassis rows in chassis_private table

Signed-off-by: Pardhakeswar Pacha <ppacha@nvidia.com>

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -375,6 +382,22 @@ func (c *ovndb) MonitorTables(jsonContext interface{}) (*libovsdb.TableUpdates,
...
+// in SB/NB DB schema and currently not in SBTablesOrder or NBTablesOrder
+func (c *ovndb) AddLatestDBSchemaTables() {
+
+	dbSchema, err := c.client.GetSchema(c.db)
+	if err != nil {
+		log.Printf("Failed to get %s db schema: (%v)", c.db, err)
+		return
+	}
+	if c.db == DBSB {
+		if _, ok := dbSchema.Tables[TableChassisPrivate]; ok {
```
Comment (girishmg) [inline client.go:395]: I think it is better to use  an array and then walk through the elements of the array, instead of directly checking against TableChassisPrivate.

As in,

var latest SBTables = []string{TableChassisPrivate}

then you can loop through these tables to check whether they are in the dbSchema.Tables.

File: client.go [hunk 2]
```diff
@@ -28,8 +28,9 @@ import (
 )
 
 type EntityType string
-const(
-	PORT_GROUP EntityType = "PORT_GROUP"
+
+const (
+	PORT_GROUP     EntityType = "PORT_GROUP"
 	LOGICAL_SWITCH EntityType = "LOICAL_SWITCH"
```
Comment (girishmg) [inline client.go:34]: s/LOICAL/LOGICAL/

File: client.go [hunk 3]
```diff
@@ -330,6 +336,7 @@ func (c *ovndb) reconnect() {
 func (c *ovndb) MonitorTables(jsonContext interface{}) (*libovsdb.TableUpdates, error) {
 	// get the table list based on the DB
 	var tables []string
+	c.AddLatestDBSchemaTables()
```
Comment (girishmg) [inline client.go:339]: L339 is inserted in-between L338-344 and the comment @L337 was explaining that. Perhaps you need to move this line to L337

Comment (Pardhakeswar) [reply client.go:339]: okay, will do the change.

File: client.go [hunk 4]
```diff
@@ -375,6 +384,24 @@ func (c *ovndb) MonitorTables(jsonContext interface{}) (*libovsdb.TableUpdates,
...
+// AddLatestSchemaTables adds new tables that needs to be monitored
+// in SB/NB DB schema and currently not in SBTablesOrder or NBTablesOrder
+func (c *ovndb) AddLatestDBSchemaTables() {
+
+	dbSchema, err := c.client.GetSchema(c.db)
+	if err != nil {
+		log.Printf("Failed to get %s db schema: (%v)", c.db, err)
+		return
+	}
+	if c.db == DBSB {
```
Comment (vtolstov) [inline client.go:396]: return early
if c.db != DBSB {
  return
}

File: chassis_private.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		}
+	}
+	return listChassisPrivate, nil
+}
+
+func (odbi *ovndb) rowToChassisPrivate(uuid string) (*ChassisPrivate, error) {
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate][uuid]
+	if !ok {
+		return nil, fmt.Errorf("Row in Chassis_Private with uuid%s not found", uuid)
```
Comment (vtolstov) [inline chassis_private.go:130]: uuid%s ?

Comment (vtolstov) [reply chassis_private.go:130]: also use lowercase error messages

**Review by vtolstov — CHANGES_REQUESTED** [review]

**Comment by Pardhakeswar** [comment]
Thanks @vtolstov for the review. Updated the PR with requested changes.

File: chassis_private.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	return listChassisPrivate, nil
+}
+
+func (odbi *ovndb) chassisPrivateGetImp(chassis string) ([]*ChassisPrivate, error) {
+	var listChassisPrivate []*ChassisPrivate
+
+	odbi.cachemutex.RLock()
+	defer odbi.cachemutex.RUnlock()
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate]
```
Comment (vtolstov) [inline chassis_private.go:108]: will be better instead of defer and cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate]
use something like this
odbi.cachemutex.RLock()
cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate]
odbi.cachemutex.RUnlock()
if !ok {
....
}

in small functions better to avoid defer usage

File: chassis_private.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+			}
+			listChassisPrivate = append(listChassisPrivate, chPrivate)
+		}
+	}
+	return listChassisPrivate, nil
+}
+
+func (odbi *ovndb) rowToChassisPrivate(uuid string) (*ChassisPrivate, error) {
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate][uuid]
```
Comment (vtolstov) [inline chassis_private.go:128]: odbi.cachemutex.RLock()
...
odbi.cachemutex.RUnlock()

**Review by vtolstov — CHANGES_REQUESTED** [review]
needs to do small changes and will be ready to merge

File: chassis_private.go (hunk 7 above)
Comment (Pardhakeswar) [reply chassis_private.go:128]: @vtolstov **rowToChassisPrivate** is being called from **ChassisPrivateGet** & **ChassisPrivateGet**
adding  odbi.cachemutex.RLock in **rowToChassisPrivate** will lead to scenario right as we already have locked **rowToChassisPrivate**?

**Comment by Pardhakeswar** [comment]
@vtolstov updated the PR.

**Review by vtolstov — APPROVED** [review]

**Comment by vtolstov** [comment]
LGTM, @noah8713 do you want to review?

File: chassis_private_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	if len(chassisPrivate) != 1 {
+		t.Fatalf("err getting chassis in ovn-sb chassis_private table, total:%v",
+			len(chassisPrivate))
+	}
+	chName := chassisPrivate[0].Name
+	t.Logf("Chassis found in chassis-private table:%+v", chName)
+
+	// deleting the row in chassis_private table
+	t.Logf("Deleting row in OVN SB DB chassis_private table:%v", chName)
+	ocmd, err = ovndbapi.ChassisDel(chName)
```
Comment (noah8713) [inline chassis_private_test.go:61]: it should be chassisprivatedel not chassisdel.

File: client.go [hunk 9]
```diff
@@ -28,9 +28,10 @@ import (
 )
 
 type EntityType string
-const(
-	PORT_GROUP EntityType = "PORT_GROUP"
-	LOGICAL_SWITCH EntityType = "LOICAL_SWITCH"
+
```
Comment (noah8713) [inline client.go:31]: fmt should be part of different commit.

File: client.go [hunk 10]
```diff
@@ -375,6 +384,25 @@ func (c *ovndb) MonitorTables(jsonContext interface{}) (*libovsdb.TableUpdates,
 	return c.client.Monitor(c.db, jsonContext, requests)
 }
 
+// AddLatestSchemaTables adds new tables that needs to be monitored
```
Comment (noah8713) [inline client.go:387]: why need AddLatestDBSchemaTables ?
Adding new table should be part of NB/SB table order as we keep adding tables on need basis. 

Client can watch on any tables on need basis

File: client.go [hunk 11]
```diff
@@ -747,3 +787,7 @@ func (c *ovndb) sbGlobalAdd(options map[string]string) (*OvnCommand, error) {
 func (c *ovndb) sbGlobalDel() (*OvnCommand, error) {
 	return c.sbGlobalDelImp()
 }
+
+func (c *ovndb) chassisPrivateAdd(name string, external_ids map[string]string) (*OvnCommand, error) {
```
Comment (noah8713) [inline client.go:791]: Please move chassisPrivateAdd naear L433 to avoid scattering api calls in different places in function

File: common.go [hunk 12]
```diff
@@ -81,3 +82,7 @@ var SBTablesOrder = []string{
 	TableEncap,
 	TableSBGlobal,
 }
+
+var LatestSBTablesOrder = []string{
```
Comment (noah8713) [inline common.go:86]: What is the intent for LatestSBTablesOrder ?TableChassisPrivate should be in SBTablesOrder

File: client.go (hunk 10 above)
Comment (girishmg) [reply client.go:387]: @noah8713 the Chassis_Private Tables is not present in all releases of OVN. It is only present in December-2020 release. So, we add that table if OVN SB driver returns that table in its schema. Hence, it needs to be dynamic.

File: common.go (hunk 12 above)
Comment (girishmg) [reply common.go:86]: The intent again is that ChassisPrivate table was only recently added. It doesn't exist on the old releases of OVN.

File: client.go (hunk 11 above)
Comment (Pardhakeswar) [reply client.go:791]: @noah8713 **chassisPrivateAdd** api is only used as helper function in unit-tests so that was the reason I added in the section of api's which are used as helpers for unit tests.

File: common.go (hunk 12 above)
Comment (noah8713) [reply common.go:86]: thanks @girishmg for details. So if new table gets added to schema everytime, we keep changing go-ovn with latest* table? I think thats not the end goal. Also our current fork runs against latest master so should be fine. Please rectify if I got it something the other way,

File: client.go (hunk 11 above)
Comment (noah8713) [reply client.go:791]: Thanks. However, any api exposed in client should follow CRUD order. Also please check as CI is failing too:
https://travis-ci.org/github/eBay/go-ovn/builds/758743285?utm_source=github_status&utm_medium=notification

```
./chassis_private_test.go:26:23: ovndbapi.chassisPrivateAdd undefined (type Client has no field or method chassisPrivateAdd)
note: module requires Go 1.12FAIL	github.com/ebay/go-ovn [build failed]
```

File: client.go (hunk 10 above)
Comment (noah8713) [reply client.go:387]: Seems then such optimization needs to be for any new table coming up. I think we add table api support for any new table being exposed in OVN. We can add a check in CRUD calls if underlying ovn doesnt have such table exposed. Also client has the flexibility to watch on tables based on their underlying ovn version. So little hesitating as to why go-ovn should worry about if table is missing in actual ovn that its connecting to.

**Comment by noah8713** [comment]
@Pardhakeswar : Thanks for your PR. Please resolve pending/closed comments that are addressed/raised by @girishmg @vtolstov  as looking at the PR shows 10s of comments all open. Would be easy to address the new pending accordingly.

File: client.go (hunk 9 above)
Comment (Pardhakeswar) [reply client.go:31]: @noah8713 these fmt changes aren't needed with this commit?

File: client.go (hunk 10 above)
Comment (girishmg) [reply client.go:387]: @noah8713 we will break existing clients connecting to old server if we blindly update the {SB|NB}TablesOrder. Other way to solve would be to always update the  {SB|NB}TablesOrder array and remove the tables that the remote OVN server doesn't support just before we call MonitorTables().

File: client.go (hunk 9 above)
Comment (noah8713) [reply client.go:31]: no

File: client.go (hunk 11 above)
Comment (noah8713) [reply client.go:791]: This still shows failure in CI. Did you push the fix?

Comment (noah8713) [reply client.go:791]: Please move this helper function with CRUD calls in line

Comment (Pardhakeswar) [reply client.go:791]: Not yet, waiting for the resolution of the discussion on LatestSBTablesOrder

File: client.go (hunk 10 above)
Comment (noah8713) [reply client.go:387]: @girishmg : Thanks. I think clients can still pin go-ovn to particular commit to support prior/onwards dec 2020 remote OVN releases. I second for the approach of doing check if new table actually exists before calling monitoring tables. I am ok if that can be incremental PR for client.go to ensure that safety check for any new future table addition/deletion in OVN and add chassisprivate to {SB|NB}TablesOrder.

**Comment by Pardhakeswar** [comment]
@noah8713 addressed the comments and updated the PR accordingly.

File: client.go (hunk 10 above)
Comment (hzhou8) [reply client.go:387]: Hi @girishmg @noah8713 thanks for the good discussion. I agree with @noah8713 that AddLatestDBSchemaTables() and LatestSBTablesOrder look weird. In fact, I think not only LatestSBTablesOrder, but NBTablesOrder/SBTablesOrder are not needed. They are used only in MonitorTables() for validating if a table is valid. In my opinion the available tables don't need to be hardcoded. I'd suggest removing NBTablesOrder/SBTablesOrder all together, but it should be done in a separate commit.

File: chassis_private_test.go [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+	t.Logf("Listing all the chassis in ovn-sb chassis_private table")
+	chassisPrivate, err := ovndbapi.ChassisPrivateList()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(chassisPrivate) != 1 {
+		t.Fatalf("err listing chassis in ovn-sb chassis_private table, total:%v",
+			len(chassisPrivate))
+	}
+	t.Logf("Single chassis found in ovn-sb chassis_private table: %+v", chassisPrivate)
```
Comment (noah8713) [inline chassis_private_test.go:50]: chassisPrivate is a list. So use chassisPrivate[0].Name

File: common.go (hunk 12 above)
Comment (noah8713) [inline common.go:86]: remove this section.

File: client.go (hunk 10 above)
Comment (noah8713) [reply client.go:387]: Sounds good. Thanks @hzhou8 . @Pardhakeswar : Please remove the add/latestdb part and just add chassisprivate as a regular table support following other tables source as a ref. @girishmg  and team please feel free to file an incremental PR to remove NB/SBTables in general as an improvement for dynamic table realization.

**Review by noah8713 — APPROVED** [review]
lgtm

File: client.go (hunk 10 above)
Comment (girishmg) [reply client.go]: @hzhou8 as I understand go-ovn does not support all the tables in SB/NB since there are no APIs to do CRUD operations against the unsupported tables. 

For example:  If a client using go-ovn requests for OVN SB's Datapath_binding, then go-ovn will request for that table to be monitored and it will get updates from the SB server, however they can't do much with those updates since there are no APIs.

My understanding was that those tables clearly said what tables are supported for CRUD operation by this library.

Comment (hzhou8) [reply client.go]: @girishmg I think the APIs should clearly tell what's supported, and when a user specifically request for monitoring a specific table, they must have clear expectation. It doesn't make much sense for a user to request a table that is not even supported by the API. Or maybe your real concern is the default behavior, i.e. when user doesn't specify which tables to monitor, falling back to monitor-all would incur some more waste?

Comment (girishmg) [reply client.go]: @hzhou8 correct the memory footprint of the clients would explode, for example, with Port_Binding and Datapth_Binding tables from OVN SB DB in a large scale cluster.

By default,  this library should monitor the tables for which there is support for CRUD operations and the end users can filter the table further from that supported list. I think the {SB|NB}TablesOrder enabled it.

Now, with ChassisPrivate table added to SBTablesOrder array and if someone upgrades to latest go-ovn library to get the recently added PortGroup and ACL APIs, then they can't use this library against their old release of OVN since it doesn't support ChassisPrivate table.

The initial PR, added ChassisPrivate table only if the SB Schema had support for it.

Comment (hzhou8) [reply client.go]: @girishmg Thanks for clarifying and now I understand your concern. So I think it makes sense to keep the NB/SBTablesOrder arrays for the sake of saving cost when users don't explicitly specify what tables are interested. While for the backward compatibility, it is more straightforward to just keep the tables in NB/SBTablesOrder matching the APIs we support and as you proposed: "remove the tables that the remote OVN server doesn't support just before we call MonitorTables()" sounds good to me.

Comment (girishmg) [reply client.go]: @hzhou8 great! @Pardhakeswar can you please take care of this? @noah8713 FYI
